api/auth_middleware.go

```diff
@@ -24,57 +34,135 @@ func getServiceAccountID(ctx context.Context) (string, bool) {
-			authorization := r.Header.Get("authorization")
-			parts := strings.Split(authorization, " ")
-			if authorization == "" || len(parts) != 2 {
-				w.WriteHeader(http.StatusUnauthorized)
+		return http.HandlerFunc(func(responseWriter http.ResponseWriter, request *http.Request) {
+			header := request.Header.Get("authorization")
+			authHeaderContents := strings.Split(header, " ")
+			authHeader := buildAuth(header, authHeaderContents[0], authHeaderContents[1])
+
+			if !authHeader.isValid() {
+				responseWriter.WriteHeader(http.StatusUnauthorized)
 				return
 			}
+
 			var ctx context.Context
-			l := middleware.GetLogger(r.Context())
-			if parts[0] == "KeyPair" {
-				keyPair := strings.Split(parts[1], ":")
-				accessKeyPairAuth, err := sasUC.WithContext(r.Context()).
-					AuthenticateKeyPair(keyPair[0], keyPair[1])
-				if err != nil {
-					l.WithError(err).Error("auth failed")
-					w.WriteHeader(http.StatusInternalServerError)
-					return
-				}
-				w.Header().Set("x-service-account-name", accessKeyPairAuth.Name)
-				ctx = context.WithValue(r.Context(), serviceAccountIDCtxKey, accessKeyPairAuth.ServiceAccountID)
-			} else if parts[0] == "Bearer" {
-				accessToken := parts[1]
-				accessTokenAuth, err := sasUC.WithContext(r.Context()).
-					AuthenticateAccessToken(accessToken)
-				if err != nil {
-					l.WithError(err).Info("auth failed")
-					if _, ok := err.(*errors.EntityNotFoundError); ok {
-						w.WriteHeader(http.StatusUnauthorized)
-						return
-					}
-					l.Error(err)
-					w.WriteHeader(http.StatusInternalServerError)
-					return
-				}
-				w.Header().Set("x-email", accessTokenAuth.Email)
-				if accessTokenAuth.AccessToken != accessToken {
-					w.Header().Set("x-access-token", accessTokenAuth.AccessToken)
-				}
-				ctx = context.WithValue(
-					r.Context(), serviceAccountIDCtxKey, accessTokenAuth.ServiceAccountID,
-				)
+			var err error
+			logger := middleware.GetLogger(request.Context())
+
+			if authHeader.isKeyPair() {
+				ctx, err = handleKeyPairAuth(request, responseWriter, authHeader, sasUC, logger)
+			} else if authHeader.isOAuth2() {
+				ctx, err = handleOAuth2TokenAuth(request, responseWriter, authHeader, sasUC, logger)
 			} else {
-				l.WithError(errors.NewInvalidAuthorizationTypeError()).Error("auth failed")
-				w.WriteHeader(http.StatusUnauthorized)
+				handleUndefinedAuthType(responseWriter, logger)
 				return
 			}
-			next.ServeHTTP(w, r.WithContext(ctx))
+
+			if err != nil {
+				return
+			}
+
+			next.ServeHTTP(responseWriter, request.WithContext(ctx))
 		})
 	}
 }
+
+func buildAuth(header string, method string, content string) authorizationHeader {
+	var authType models.AuthenticationType
+
+	if strings.EqualFold(method, keyPairHeader) {
+		authType = models.AuthenticationTypes.KeyPair
+	} else if strings.EqualFold(method, bearerTokenHeader) {
+		authType = models.AuthenticationTypes.OAuth2
+	} else {
+		authType = ""
```

Comment:
We can receive requests with any authorisation method, even ones that we not support. By setting it as `""` it will fail later when we check if the request uses OAuth2 or KeyPair.

Maybe we could add a type for this inside `models.AuthenticationTypes`? For instance, `models.AuthenticationTypes.Unsupported` or `models.AuthenticationTypes.Unknown` 🤔 💭

Reply:
got it. I would take the Unknown route, in order to keep discretising our options.